### ISSUE TITLE
Redis DS performance slower than JSON

### ISSUE THREAD

**Issue by awesomenode** [sysrepo/sysrepo#3426]
Hello!

I was evaluating the Redis Datastore plugin using the sr_perf performance tests and got the following results (green and red were modified to + and -):
> UpdateCTestConfiguration  from :/sysrepo/build/DartConfiguration.tcl
> UpdateCTestConfiguration  from :/sysrepo/build/DartConfiguration.tcl
> Test project /sysrepo/build
> Constructing a list of tests
> Done constructing a list of tests
> Updating test list for fixtures
> Added cleanup test tests_clear_repos required by fixture tests_cleanup
> Added cleanup test tests_clear_shm required by fixture tests_cleanup
> Added 2 tests to meet fixture requirements
> Checking test dependency graph...
> Checking test dependency graph end
> test 4
>     Start 4: sr_perf_1000_10
> 
> 4: Test command: /sysrepo/build/tests/sr_perf "1000" "10"
> 4: Environment variables: 
> 4:  SYSREPO_REPOSITORY_PATH=/sysrepo/build/test_repositories/sr_perf
> 4:  SYSREPO_SHM_PREFIX=_tests_sr_sr_perf
> 4: Test timeout computed to be: 10000000
> 4: 
> 4: | Options
> 4: 
> 4:   Data set size      : 1000
> 4:   Each test executed : 10 times
> 4: 
> 4: 
> 4: | Performance tests
> 4: 
> 4:   JSON DS file
> 4:  ___________________________________________________________________ 
> 4: |                                                                   |
> 4: | test name                          |         time |    comparison |
> 4: |___________________________________________________________________|
> 4: |                                                                   |
> 4: | get tree ......................... |   0.002337 s |       1.000 x |
> 4: | get item ......................... |   0.002484 s |       1.000 x |
> 4: | get tree hash .................... |   0.001682 s |       1.000 x |
> 4: | get tree hash cached ............. |   0.000016 s |       1.000 x |
> 4: | get user ordered tree ............ |   0.001240 s |       1.000 x |
> 4: | get oper tree .................... |   0.004691 s |       1.000 x |
> 4: | create batch ..................... |   0.014343 s |       1.000 x |
> 4: | create user ordered items ........ |   0.015998 s |       1.000 x |
> 4: | create all items ................. |   0.023072 s |       1.000 x |
> 4: | create all items oper ............ |   0.035199 s |       1.000 x |
> 4: | remove all items ................. |   0.000332 s |       1.000 x |
> 4: | remove all items cached .......... |   0.000283 s |       1.000 x |
> 4: | remove whole subtree ............. |   0.000400 s |       1.000 x |
> 4: | remove whole subtree cached ...... |   0.000353 s |       1.000 x |
> 4: | create an item ................... |   0.011295 s |       1.000 x |
> 4: | create an item cached ............ |   0.010626 s |       1.000 x |
> 4: | create an item oper .............. |   0.024746 s |       1.000 x |
> 4: | modify an item ................... |   0.002592 s |       1.000 x |
> 4: | modify an item cached ............ |   0.001533 s |       1.000 x |
> 4: | remove an item ................... |   0.002566 s |       1.000 x |
> 4: | remove an item cached ............ |   0.001507 s |       1.000 x |
> 4: |___________________________________________________________________|
> 4: 
> 4: 
> 4:   REDIS DS
> 4:  ___________________________________________________________________ 
> 4: |                                                                   |
> 4: | test name                          |         time |    comparison |
> 4: |___________________________________________________________________|
> 4: |                                                                   |
> 4: | get tree ......................... |   0.026150 s |    - 11.189 x |
> 4: | get item ......................... |   0.020679 s |     - 8.324 x |
> 4: | get tree hash .................... |   0.001196 s |     + 1.406 x |
> 4: | get tree hash cached ............. |   0.000770 s |    - 48.125 x |
> 4: | get user ordered tree ............ |   0.027212 s |    - 21.945 x |
> 4: | get oper tree .................... |   0.008245 s |     - 1.757 x |
> 4: | create batch ..................... |   0.056427 s |     - 3.934 x |
> 4: | create user ordered items ........ |   0.563826 s |    - 35.243 x |
> 4: | create all items ................. |   0.039440 s |     - 1.709 x |
> 4: | create all items oper ............ |   0.089823 s |     - 2.551 x |
> 4: | remove all items ................. |   0.004348 s |    - 13.096 x |
> 4: | remove all items cached .......... |   0.002610 s |     - 9.222 x |
> 4: | remove whole subtree ............. |   0.005553 s |    - 13.882 x |
> 4: | remove whole subtree cached ...... |   0.001713 s |     - 4.852 x |
> 4: | create an item ................... |   0.024816 s |     - 2.197 x |
> 4: | create an item cached ............ |   0.003018 s |     + 3.520 x |
> 4: | create an item oper .............. |   0.100182 s |     - 4.048 x |
> 4: | modify an item ................... |   0.019461 s |     - 7.508 x |
> 4: | modify an item cached ............ |   0.001185 s |     + 1.293 x |
> 4: | remove an item ................... |   0.020534 s |     - 8.002 x |
> 4: | remove an item cached ............ |   0.001057 s |     + 1.425 x |
> 4: |___________________________________________________________________|
> 4: 
> 4: 
> 4: All comparisons refer to how many times faster (+) or slower (-) the current plugin is compared to the first plugin.
> 4: 
> 4:  
> 1/3 Test #4: sr_perf_1000_10 ..................   Passed   22.54 sec
> test 2
>     Start 2: tests_clear_repos
> 
> 2: Test command: /usr/bin/rm "-rf" "/sysrepo/build/test_repositories"
> 2: Test timeout computed to be: 10000000
> 2/3 Test #2: tests_clear_repos ................   Passed    0.00 sec
> test 3
>     Start 3: tests_clear_shm
> 
> 3: Test command: /usr/bin/find "/dev/shm/" "-maxdepth" "1" "-readable" "-name" "_tests_sr_*" "-delete"
> 3: Test timeout computed to be: 10000000
> 3/3 Test #3: tests_clear_shm ..................   Passed    0.00 sec
> 
> The following tests passed:
>         sr_perf_1000_10
>         tests_clear_repos
>         tests_clear_shm
> 
> 100% tests passed, 0 tests failed out of 3
> 
> Total Test time (real) =  22.56 sec

I was using the setup from the attached dockerfile, making no modifications to the default redis or sysrepo configurations. Is the slower performance in most test cases expected, or is there a way to speed them up with configuration changes?
[Dockerfile.zip](https://github.com/user-attachments/files/17226597/Dockerfile.zip)

**Comment by michalvasko** [comment]
I was hoping the student responsible for these plugins will answer you but it seems it will still take some time, so I will try responding best I can. Firstly, these plugins are meant for really large datastores, 10k items and more. So the more data you have, the faster should these DB plugins be compared to JSON. But operations on the whole datastore (create/delete all items, ...) will always be slower with DB, that is what JSON is technically optimized for but in practice it should be rare, you almost always do smaller changes. Lastly, the student was still optimizing these plugins and I am not sure that everything has already been committed. So I wanted to get his numbers to compare them with yours to make sure there are no unexpected differences.

**Comment by Humblesaw** [comment]
Yes, the performance is very bad. Most test cases work with the data of the whole YANG model, however this is not what the database plugins were designed for. They should speed up the storing/loading of low amounts of data in a high-volume data environment. This is only true for cached connections (because without cache you first need to load the whole data tree which is slower in the database). However even then the numbers are not great. I took your Dockerfile, adjusted it and ran the performance tests on the latest libyang and sysrepo versions, you should be able to see it in the devel branch of the sysrepo. There is also a thorough explanation of the results in DB_PLG_PERF.md.
